[WIP] Replicate RAJAPerf Experiments

michaelmckinsey1 opened the pull request:
## Description

Enable replication of RAJAPerf experiments from recent Jan 2025 paper. These experiments were not ran in Benchpark, but were built with cmake & build scripts and ran manually.

- [x] Enable passing additional command line arguments to `raja-perf` binary
- [ ] Caliper configs we need to be able to use:
   - [x] time
   - [ ] topdown
   - [ ] gpu-side calls (cuda-activity-profile)
   - [ ] ncu 
- [ ] Replicate Datasets:
   - [x] V100
   - [ ] MI250x
   - [ ] MI300a
   - [ ] A100
   - [ ] GH200

## Adding/modifying a benchmark (docs: [Adding a Benchmark](https://software.llnl.gov/benchpark/add-a-benchmark.html))

- [x] Modify `experiments/raja-perf/experiment.py`
- [x] Modify `repo/raja-perf/application.py`


## V100 setup reproducer

```
benchpark system init --dest=lassen llnl-sierra
benchpark experiment init --dest=raja-perf raja-perf+cuda+strong~single_node caliper=time version=2024.07.0 scaling-factor=4 scaling-iterations=2 total_size=33554432 repfact=5 variants="Base_Seq-RAJA_Seq" tunings="default-library"
benchpark setup ./raja-perf/ ./lassen/ wkp/
```
![image](https://github.com/user-attachments/assets/6a144620-d8ea-47a5-8806-099e306309c9)
(text values shown on figure are raw values, not speedup)

Commit:
Enable passing problem size to experiment

Commit:
Expose more arguments to enable exact match of parameters

Commit:
Undo variant/tuning change that messed up experiment_run_dir

michaelmckinsey1 commented on repo/raja-perf/application.py:
```diff
@@ -10,18 +10,68 @@
+    )
+
+    # executable(
+    #     "run",
+    #     template=["{execute}" + " --size {size}" + " --repfact {repfact}" + " {additional_args}"],
+    #     use_mpi=True
+    # )
+
+    workload("suite", executables=["run"])
+
+    # workload_variable(
+    #     "additional_args",
+    #     default="--variants {variants} --tunings {tunings}",
+    #     workloads=["suite"],
+    #     description="",
+    # )
+
+    # workload_variable("variants", default="", workloads=["suite"], description="")
 
-    executable('run', 'raja-perf.exe', use_mpi=True)
+    # workload_variable("tunings", default="", workloads=["suite"], description="")
```
I would like to have the option for `--variants` and `--tunings` instead of having to manually edit the `execute_experiment` files. The problem with this code is that when no argument is provided, the arguments are still injected, e.g. `raja-perf.exe --repfact 5 ... --variants --tunings`, when ideally they should not appear if not provided `raja-perf.exe --repfact 5 ...`

This can be done by checking in `experiment.py`, but that code results in the arguments appearing into `experiment_run_dir` (e.g. `...raja-perf.exe/raja-perf_suite_strong_1_33554432_5 --variants RAJA_CUDA --tunings block_256 ...`) since `execute` is an experiment variable.

Commit:
Enable setting ranks as variant

pearce8 commented on experiments/raja-perf/experiment.py:
```diff
@@ -56,6 +82,23 @@ def compute_applications_section(self):
+
+        self.add_experiment_variable("repfact", self.spec.variants["repfact"][0], True)
+
+        # rajaperf_variants = self.spec.variants["variants"][0].replace("-", " ")
```
Should the commented out code be removed?

michaelmckinsey1 replied:
It has been removed.

pearce8 commented on repo/raja-perf/application.py:
```diff
@@ -10,18 +10,68 @@
+
+    workload("suite", executables=["run"])
+
+    # workload_variable(
```
Should the commented out code be removed?

michaelmckinsey1 replied:
It has been removed.

pearce8 commented on repo/raja-perf/application.py:
```diff
@@ -10,18 +10,68 @@
+        use_mpi=True
+    )
+
+    # executable(
```
Should the commented out code be removed?

michaelmckinsey1 replied:
It has been removed.

Commit:
Add support for variants and tunings and remove commented code

michaelmckinsey1 committed:
Undo n_ranks variant

michaelmckinsey1 committed:
Update scaling implementation

michaelmckinsey1 committed:
Add branch for testing

pearce8 reviewed (changes requested):
@michaelmckinsey1 where does this stand?  If you still want to merge it, please update it.  Otherwise, please close this PR.

michaelmckinsey1 commented:
Depends on #750

michaelmckinsey1 commented:
> @michaelmckinsey1 where does this stand? If you still want to merge it, please update it. Otherwise, please close this PR.

#750 has some fixes to the rajaperf experiment which supersede the changes here. I need to reconcile the two and then do some testing on the different systems.

michaelmckinsey1 commented:
This PR has become very outdated and partially superseded by #750 and #829. There are still unaddressed issues that I've summarized in https://github.com/LLNL/benchpark/issues/843, and I plan to create new PRs to address these.
